### DISCUSSION TITLE
What could be the reason for "Table 'xcom' is already defined for this MetaData instance" error?

### DISCUSSION THREAD

**Discussion by makrushin-evgenii** [apache/airflow#17221] [Q&A]
I get this error when using my custom XCom. The error appears only in the production environment, I could not reproduce it locally
```
sqlalchemy.exc.InvalidRequestError: Table 'xcom' is already defined for this MetaData instance.  Specify 'extend_existing=True' to redefine options and columns on an existing Table object.
[2021-07-26 07:11:51,139] {plugins_manager.py:232} ERROR - Failed to import plugin /opt/airflow/dags/repo/plugins/db_worker_plugin.py
```

Can you suggest a possible cause and solution to this problem? The solutions I have found either require changes to the Airflow code or removal of tables in the Metabase, both of which don't work well for me.

**Comment by potiuk** [comment]
I am afraid It's something wrong in your custom plugin. So it's impossible to say what the plugin does wrongly without seing it. You need to find authors of the plugin in your company or analyse it's code and fix it.

Note that it is not recommended to keep custom plugins in your dag folder. It's better if you keep them separately and install in your airflow plugins via the same mechanisms as any other python packages: https://airflow.apache.org/docs/apache-airflow/stable/plugins.html#plugins-as-python-packages

**Reply by makrushin-evgenii** [reply]
Thank you for your answer. If I understand correctly, the problem may not be with the plugin itself, but with the way it is installed

Must I put the source code of all the plugins in a separate python package, and install it at the image building stage? Or it shouldn't make any difference if the problem is with the plugin itself

Source code of the plugin:
```python
from airflow.models import BaseOperator
from airflow.plugins_manager import AirflowPlugin
from airflow.providers.microsoft.mssql.operators.mssql import MsSqlHook
from math import ceil

class DBWorker(BaseOperator):
    def __init__(self):
        self.hook = MsSqlHook('timepad_vm_um_dwh')

    def recreate_table_by_df(self, name, df):
         self.hook.run('create table timepad.{0} ({1})'.format(name, self._create_sql_str_from_dtypes(df)), autocommit=True)

    @staticmethod
    def _create_sql_str_from_dtypes(df):
        d = df.dtypes.apply(lambda x: x.name).to_dict()
        mapping = {'object': 'varchar(2000)', 'int64': 'int', 'float64': 'float', 'bool': 'bit',
                   'datetime64[ns]': 'datetime', 'datetime64[ns, pytz.FixedOffset(180)]': 'datetime',
                   'datetime64[ns, UTC]': 'datetime'}

        return ', '.join([k + ' ' + mapping[v] for k, v in d.items()])

    def insert_rows(self,table,df):
        sql_head = 'INSERT INTO timepad.{0} ({1}) values '.format(table, ','.join(df.columns))
        for part in range(ceil(len(df.index)/100)):
            values = ''
            for i, row in df[part*100:part*100 + 100].iterrows():
                values = values + '('
                for col in df.columns:
                    values = values + "'" + str(row[col]) + "',"
                values = values[:-1] + '),'
            sql = sql_head + values[:-1] + ';'
            self.hook.run(sql, autocommit=True)

    def recreate_indexes(self):
        self.hook.run('exec timepad.p_step1_alter_id_cols;', autocommit=True)
        self.hook.run('exec timepad.p_step2_add_pk;', autocommit=True)
        self.hook.run('exec timepad.p_step3_add_fk;', autocommit=True)

    def drop_all_tables(self):
        self.hook.run('''DROP table if exists timepad.questions;
                         DROP table if exists timepad.tickets;
                         DROP table if exists timepad.company_payment_requests;
                         DROP table if exists timepad.companies;
                         DROP table if exists timepad.invoices;
                         DROP table if exists timepad.orders;
                         DROP table if exists timepad.events;
                         DROP table if exists timepad.orgs;'''
                      , autocommit=True)

class DBWorkerPlugin(AirflowPlugin):
    name = "db_worker_plugin"
    operators = [DBWorker]
```

**Reply by potiuk** [reply]
It does not matter how it is installed. But it's your plugin, so I am afraid you need to understand what it does to solve the problem :)

**Comment by diman82** [comment]
@makrushin-evgenii Go to the very beginning of the log, and you may find the true root cause of the problem.
In my case, I got the very same error message, but beginning of the log revealed, that true reason for the failing startup was an incorrect path:
`[[34m2022-07-26 09:46:47,566[0m] {[34mconfiguration.py:[0m676} ERROR[0m - No module named 'include.xcom_wrapper'[0m
[[34m2022-07-26 09:46:47,566[0m] {[34mplugins_manager.py:[0m264} ERROR[0m - Failed to import plugin /usr/local/airflow/plugins/macros/custom_macros.py[0m
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/airflow/configuration.py", line 674, in getimport
    return import_string(full_qualified_path)
  File "/usr/local/lib/python3.9/site-packages/airflow/utils/module_loading.py", line 32, in import_string
    module = import_module(module_path)
  File "/usr/local/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 984, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'include.xcom_wrapper'

....

  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/deprecations.py", line 309, in warned
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 593, in __new__
    raise exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: Table 'xcom' is already defined for this MetaData instance.  Specify 'extend_existing=True' to redefine options and columns on an existing Table object.
 `
